File: src/main/java/at/bitfire/ical4android/AndroidTask.java
/*
 * Copyright (c) 2013 – 2015 Ricki Hirner (bitfire web engineering).
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE.  See the GNU General Public License for more details.
 */

package at.bitfire.ical4android;

import android.content.ContentProviderOperation;
import android.content.ContentProviderOperation.Builder;
import android.content.ContentUris;
import android.content.ContentValues;
import android.database.Cursor;
import android.database.DatabaseUtils;
import android.net.Uri;
import android.os.RemoteException;
import android.text.TextUtils;
import android.util.Log;

import net.fortuna.ical4j.model.Date;
import net.fortuna.ical4j.model.DateTime;
import net.fortuna.ical4j.model.Dur;
import net.fortuna.ical4j.model.TimeZone;
import net.fortuna.ical4j.model.property.Clazz;
import net.fortuna.ical4j.model.property.Completed;
import net.fortuna.ical4j.model.property.DtStart;
import net.fortuna.ical4j.model.property.Due;
import net.fortuna.ical4j.model.property.Duration;
import net.fortuna.ical4j.model.property.ExDate;
import net.fortuna.ical4j.model.property.Geo;
import net.fortuna.ical4j.model.property.Organizer;
import net.fortuna.ical4j.model.property.RDate;
import net.fortuna.ical4j.model.property.RRule;
import net.fortuna.ical4j.model.property.Status;

import org.apache.commons.lang3.StringUtils;
import org.dmfs.provider.tasks.TaskContract.Tasks;

import java.io.FileNotFoundException;
import java.net.URI;
import java.net.URISyntaxException;
import java.text.ParseException;

import lombok.Cleanup;
import lombok.Getter;

public abstract class AndroidTask {
    private static final String TAG = "ical4android.Task";

    final protected AndroidTaskList taskList;

    @Getter
    protected Long id;

    protected Task task;


    protected AndroidTask(AndroidTaskList taskList, long id) {
        this.taskList = taskList;
        this.id = id;
    }

    protected AndroidTask(AndroidTaskList taskList, Task task) {
        this.taskList = taskList;
        this.task = task;
    }


    public Task getTask() throws FileNotFoundException, CalendarStorageException {
        if (task != null)
            return task;

        try {
            task = new Task();
            @Cleanup final Cursor cursor = taskList.provider.client.query(taskSyncURI(), null, null, null, null);
            if (cursor != null && cursor.moveToFirst()) {
                ContentValues values = new ContentValues(cursor.getColumnCount());
                DatabaseUtils.cursorRowToContentValues(cursor, values);
                populateTask(values);
            } else
                throw new FileNotFoundException("Couldn't load details of task #" + id);
            return task;
        } catch(RemoteException e) {
            throw new CalendarStorageException("Couldn't read locally stored event", e);
        } catch(ParseException e) {
            throw new CalendarStorageException("Couldn't parse locally stored event", e);
        }
    }

    protected void populateTask(ContentValues values) throws FileNotFoundException, RemoteException, ParseException {
        task.uid = values.getAsString(Tasks._UID);
        task.summary = values.getAsString(Tasks.TITLE);
        task.location = values.getAsString(Tasks.LOCATION);

        if (values.containsKey(Tasks.GEO)) {
            String geo = values.getAsString(Tasks.GEO);
            if (geo != null)
                task.geoPosition = new Geo(geo);
        };

        task.description = StringUtils.stripToNull(values.getAsString(Tasks.DESCRIPTION));
        task.url = StringUtils.stripToNull(values.getAsString(Tasks.URL));

        String organizer = values.getAsString(Tasks.ORGANIZER);
        if (!TextUtils.isEmpty(organizer))
            try {
                task.organizer = new Organizer("mailto:" + values.getAsString(Tasks.ORGANIZER));
            } catch (URISyntaxException e) {
                Log.w(TAG, "Invalid ORGANIZER email", e);
            }

        Integer priority = values.getAsInteger(Tasks.PRIORITY);
        if (priority != null)
            task.priority = priority;

        Integer classification = values.getAsInteger(Tasks.CLASSIFICATION);
        if (classification != null)
            switch (classification) {
                case Tasks.CLASSIFICATION_PUBLIC:
                    task.classification = Clazz.PUBLIC;
                    break;
                case Tasks.CLASSIFICATION_CONFIDENTIAL:
                    task.classification = Clazz.CONFIDENTIAL;
                    break;
                default:
                    task.classification = Clazz.PRIVATE;
            }

        Long completed = values.getAsLong(Tasks.COMPLETED);
        if (completed != null)
            // COMPLETED must always be a DATE-TIME
            task.completedAt = new Completed(new DateTime(completed));

        Integer percentComplete = values.getAsInteger(Tasks.PERCENT_COMPLETE);
        if (percentComplete != null)
            task.percentComplete = percentComplete;

        Integer status = values.getAsInteger(Tasks.STATUS);
        if (status != null)
            switch (status) {
                case Tasks.STATUS_IN_PROCESS:
                    task.status = Status.VTODO_IN_PROCESS;
                    break;
                case Tasks.STATUS_COMPLETED:
                    task.status = Status.VTODO_COMPLETED;
                    break;
                case Tasks.STATUS_CANCELLED:
                    task.status = Status.VTODO_CANCELLED;
                    break;
                default:
                    task.status = Status.VTODO_NEEDS_ACTION;
            }

        boolean allDay = false;
        if (values.getAsInteger(Tasks.IS_ALLDAY) != null)
            allDay = values.getAsInteger(Tasks.IS_ALLDAY) != 0;

        String tzID = values.getAsString(Tasks.TZ);
        TimeZone tz = (tzID != null) ? DateUtils.tzRegistry.getTimeZone(tzID) : null;

        Long createdAt = values.getAsLong(Tasks.CREATED);
        if (createdAt != null)
            task.createdAt = createdAt;

        Long lastModified = values.getAsLong(Tasks.LAST_MODIFIED);
        if (lastModified != null)
            task.lastModified = lastModified;

        Long dtStart = values.getAsLong(Tasks.DTSTART);
        if (dtStart != null) {
            long ts = dtStart;

            Date dt;
            if (allDay)
                dt = new Date(ts);
            else {
                dt = new DateTime(ts);
                if (tz != null)
                    ((DateTime)dt).setTimeZone(tz);
            }
            task.dtStart = new DtStart(dt);
        }

        Long due = values.getAsLong(Tasks.DUE);
        if (due != null) {
            long ts = due;

            Date dt;
            if (allDay)
                dt = new Date(ts);
            else {
                dt = new DateTime(ts);
                if (tz != null)
                    ((DateTime)dt).setTimeZone(tz);
            }
            task.due = new Due(dt);
        }

        String duration = values.getAsString(Tasks.DURATION);
        if (duration != null)
            task.duration = new Duration(new Dur(duration));

        String rDate = values.getAsString(Tasks.RDATE);
        if (rDate != null)
            task.getRDates().add((RDate)DateUtils.androidStringToRecurrenceSet(rDate, RDate.class, allDay));

        String exDate = values.getAsString(Tasks.EXDATE);
        if (exDate != null)
            task.getExDates().add((ExDate)DateUtils.androidStringToRecurrenceSet(exDate, ExDate.class, allDay));

        String rRule = values.getAsString(Tasks.RRULE);
        if (rRule != null)
            task.rRule = new RRule(rRule);
    }


    public Uri add() throws CalendarStorageException {
        BatchOperation batch = new BatchOperation(taskList.provider.client);
        Builder builder = ContentProviderOperation.newInsert(taskList.syncAdapterURI(taskList.provider.tasksUri()));
        buildTask(builder, false);
        batch.enqueue(builder.build());
        batch.commit();
        return batch.getResult(0).uri;
    }

    public void update(Task task) throws CalendarStorageException {
        this.task = task;

        BatchOperation batch = new BatchOperation(taskList.provider.client);
        Builder builder = ContentProviderOperation.newUpdate(taskSyncURI());
        buildTask(builder, true);
        batch.enqueue(builder.build());
        batch.commit();
    }

    public int delete() throws CalendarStorageException {
        try {
            return taskList.provider.client.delete(taskSyncURI(), null, null);
        } catch (RemoteException e) {
            throw new CalendarStorageException("Couldn't delete event", e);
        }
    }

    protected void buildTask(Builder builder, boolean update) {
        if (!update)
            builder .withValue(Tasks.LIST_ID, taskList.getId())
                    .withValue(Tasks._DIRTY, 0);

        builder
                //.withValue(Tasks._UID, task.uid)          // not available in F-Droid OpenTasks version yet (15 Oct 2015)
                .withValue(Tasks.TITLE, task.summary)
                .withValue(Tasks.LOCATION, task.location);

        if (task.geoPosition != null)
                builder.withValue(Tasks.GEO, task.geoPosition.getValue());

        builder .withValue(Tasks.DESCRIPTION, task.description)
                .withValue(Tasks.URL, task.url);

        if (task.organizer != null)
            try {
                URI organizer = new URI(task.organizer.getValue());
                if ("mailto".equals(organizer.getScheme()))
                    builder.withValue(Tasks.ORGANIZER, organizer.getSchemeSpecificPart());
                else
                    Log.w(TAG, "Found non-mailto ORGANIZER URI, ignoring");
            } catch (URISyntaxException e) {
                e.printStackTrace();
            }

        builder.withValue(Tasks.PRIORITY, task.priority);

        if (task.classification != null) {
            int classCode = Tasks.CLASSIFICATION_PRIVATE;
            if (task.classification == Clazz.PUBLIC)
                classCode = Tasks.CLASSIFICATION_PUBLIC;
            else if (task.classification == Clazz.CONFIDENTIAL)
                classCode = Tasks.CLASSIFICATION_CONFIDENTIAL;
            builder.withValue(Tasks.CLASSIFICATION, classCode);
        }

        if (task.completedAt != null) {
            // COMPLETED must always be a DATE-TIME
            builder .withValue(Tasks.COMPLETED, task.completedAt.getDateTime().getTime())
                    .withValue(Tasks.COMPLETED_IS_ALLDAY, 0);
        }
        builder.withValue(Tasks.PERCENT_COMPLETE, task.percentComplete);

        int statusCode = Tasks.STATUS_DEFAULT;
        if (task.status != null) {
            if (task.status == Status.VTODO_NEEDS_ACTION)
                statusCode = Tasks.STATUS_NEEDS_ACTION;
            else if (task.status == Status.VTODO_IN_PROCESS)
                statusCode = Tasks.STATUS_IN_PROCESS;
            else if (task.status == Status.VTODO_COMPLETED)
                statusCode = Tasks.STATUS_COMPLETED;
            else if (task.status == Status.VTODO_CANCELLED)
                statusCode = Tasks.STATUS_CANCELLED;
        }
        builder.withValue(Tasks.STATUS, statusCode);

        final boolean allDay = task.isAllDay();
        if (allDay)
            builder.withValue(Tasks.IS_ALLDAY, 1);
        else {
            builder.withValue(Tasks.IS_ALLDAY, 0);

            java.util.TimeZone tz = task.getTimeZone();
            builder.withValue(Tasks.TZ, tz.getID());
        }

        if (task.createdAt != null)
            builder.withValue(Tasks.CREATED, task.createdAt);
        if (task.lastModified != null)
            builder.withValue(Tasks.LAST_MODIFIED, task.lastModified);

        if (task.dtStart != null)
            builder.withValue(Tasks.DTSTART, task.dtStart.getDate().getTime());
        if (task.due != null)
            builder.withValue(Tasks.DUE, task.due.getDate().getTime());
        if (task.duration != null)
            builder.withValue(Tasks.DURATION, task.duration.getValue());

        if (!task.getRDates().isEmpty())
            try {
                builder.withValue(Tasks.RDATE, DateUtils.recurrenceSetsToAndroidString(task.getRDates(), allDay));
            } catch (ParseException e) {
                Log.e(TAG, "Couldn't parse RDate(s)", e);
            }
        if (!task.getExDates().isEmpty())
            try {
                builder.withValue(Tasks.EXDATE, DateUtils.recurrenceSetsToAndroidString(task.getExDates(), allDay));
            } catch (ParseException e) {
                Log.e(TAG, "Couldn't parse ExDate(s)", e);
            }
        if (task.rRule != null)
            builder.withValue(Tasks.EXDATE, task.rRule.getValue());
    }


    protected Uri tasksSyncURI() {
        return taskList.syncAdapterURI(taskList.provider.tasksUri());
    }

    protected Uri taskSyncURI() {
        if (id == null)
            throw new IllegalStateException("Task doesn't have an ID yet");
        return taskList.syncAdapterURI(ContentUris.withAppendedId(taskList.provider.tasksUri(), id));
    }

}

File: src/main/java/at/bitfire/ical4android/AndroidTask.kt
/***************************************************************************************************
 * Copyright © All Contributors. See LICENSE and AUTHORS in the root directory for details.
 **************************************************************************************************/

package at.bitfire.ical4android

import android.content.ContentUris
import android.content.ContentValues
import android.net.Uri
import android.os.RemoteException
import androidx.annotation.CallSuper
import at.bitfire.ical4android.BatchOperation.CpoBuilder
import at.bitfire.ical4android.util.AndroidTimeUtils
import at.bitfire.ical4android.util.DateUtils
import at.bitfire.ical4android.util.MiscUtils
import at.bitfire.ical4android.util.MiscUtils.CursorHelper.toValues
import net.fortuna.ical4j.model.*
import net.fortuna.ical4j.model.Date
import net.fortuna.ical4j.model.TimeZone
import net.fortuna.ical4j.model.component.VAlarm
import net.fortuna.ical4j.model.parameter.Email
import net.fortuna.ical4j.model.parameter.RelType
import net.fortuna.ical4j.model.parameter.Related
import net.fortuna.ical4j.model.property.*
import net.fortuna.ical4j.util.TimeZones
import org.dmfs.tasks.contract.TaskContract.Properties
import org.dmfs.tasks.contract.TaskContract.Property.*
import org.dmfs.tasks.contract.TaskContract.Tasks
import java.io.FileNotFoundException
import java.net.URISyntaxException
import java.time.ZoneId
import java.util.*
import java.util.logging.Level

/**
 * Stores and retrieves VTODO iCalendar objects (represented as [Task]s) to/from the
 * OpenTasks provider.
 *
 * Extend this class to process specific fields of the task.
 *
 * The SEQUENCE field is stored in [Tasks.SYNC_VERSION], so don't use [Tasks.SYNC_VERSION]
 * for anything else.
 *
 */
abstract class AndroidTask(
        val taskList: AndroidTaskList<AndroidTask>
) {

    companion object {
        const val UNKNOWN_PROPERTY_DATA = Properties.DATA0

        val utcTimeZone by lazy { DateUtils.ical4jTimeZone(TimeZones.UTC_ID) }
    }

    var id: Long? = null


    constructor(taskList: AndroidTaskList<AndroidTask>, values: ContentValues): this(taskList) {
        id = values.getAsLong(Tasks._ID)
    }

    constructor(taskList: AndroidTaskList<AndroidTask>, task: Task): this(taskList) {
        this.task = task
    }


    var task: Task? = null
        /**
         * This getter returns the full task data, either from [task] or, if [task] is null, by reading task
         * number [id] from the task provider
         * @throws IllegalArgumentException if task has not been saved yet
         * @throws FileNotFoundException if there's no task with [id] in the task provider
         * @throws RemoteException on task provider errors
         */
        get() {
            if (field != null)
                return field
            val id = requireNotNull(id)

            try {
                val client = taskList.provider.client
                client.query(taskSyncURI(true), null, null, null, null)?.use { cursor ->
                    if (cursor.moveToFirst()) {
                        // create new Task which will be populated
                        val newTask = Task()
                        field = newTask

                        val values = cursor.toValues(true)
                        Ical4Android.log.log(Level.FINER, "Found task", values)
                        populateTask(values)

                        if (values.containsKey(Properties.PROPERTY_ID)) {
                            // process the first property, which is combined with the task row
                            populateProperty(values)

                            while (cursor.moveToNext()) {
                                // process the other properties
                                populateProperty(cursor.toValues(true))
                            }
                        }

                        // Special case: parent_id set, but no matching parent Relation row (like given by aCalendar+)
                        val relatedToList = newTask.relatedTo
                        values.getAsLong(Tasks.PARENT_ID)?.let { parentId ->
                            val hasParentRelation = relatedToList.any { relatedTo ->
                                val relatedType = relatedTo.getParameter<RelType>(Parameter.RELTYPE)
                                relatedType == RelType.PARENT || relatedType == null /* RelType.PARENT is the default value */
                            }
                            if (!hasParentRelation) {
                                // get UID of parent task
                                val parentContentUri = ContentUris.withAppendedId(taskList.tasksSyncUri(), parentId)
                                client.query(parentContentUri, arrayOf(Tasks._UID), null, null, null)?.use { cursor ->
                                    if (cursor.moveToNext()) {
                                        // add RelatedTo for parent task
                                        relatedToList += RelatedTo(cursor.getString(0))
                                    }
                                }
                            }
                        }

                        field = newTask
                        return newTask
                    }
                }
            } catch (e: Exception) {
                /* Populating event has been interrupted by an exception, so we reset the event to
                avoid an inconsistent state. This also ensures that the exception will be thrown
                again on the next get() call. */
                field = null
                throw e
            }
            throw FileNotFoundException("Couldn't find task #$id")
        }

    @CallSuper
    protected open fun populateTask(values: ContentValues) {
        val task = requireNotNull(task)

        task.uid = values.getAsString(Tasks._UID)
        task.sequence = values.getAsInteger(Tasks.SYNC_VERSION)
        task.summary = values.getAsString(Tasks.TITLE)
        task.location = values.getAsString(Tasks.LOCATION)

        values.getAsString(Tasks.GEO)?.let { geo ->
            val (lng, lat) = geo.split(',')
            try {
                task.geoPosition = Geo(lat.toBigDecimal(), lng.toBigDecimal())
            } catch (e: NumberFormatException) {
                Ical4Android.log.warning("Invalid GEO value: $geo")
            }
        }

        task.description = values.getAsString(Tasks.DESCRIPTION)
        task.color = values.getAsInteger(Tasks.TASK_COLOR)
        task.url = values.getAsString(Tasks.URL)

        values.getAsString(Tasks.ORGANIZER)?.let {
            try {
                task.organizer = Organizer("mailto:$it")
            } catch(e: URISyntaxException) {
                Ical4Android.log.log(Level.WARNING, "Invalid ORGANIZER email", e)
            }
        }

        values.getAsInteger(Tasks.PRIORITY)?.let { task.priority = it }

        task.classification = when (values.getAsInteger(Tasks.CLASSIFICATION)) {
            Tasks.CLASSIFICATION_PUBLIC ->       Clazz.PUBLIC
            Tasks.CLASSIFICATION_PRIVATE ->      Clazz.PRIVATE
            Tasks.CLASSIFICATION_CONFIDENTIAL -> Clazz.CONFIDENTIAL
            else ->                              null
        }

        values.getAsLong(Tasks.COMPLETED)?.let { task.completedAt = Completed(DateTime(it)) }
        values.getAsInteger(Tasks.PERCENT_COMPLETE)?.let { task.percentComplete = it }

        task.status = when (values.getAsInteger(Tasks.STATUS)) {
            Tasks.STATUS_IN_PROCESS -> Status.VTODO_IN_PROCESS
            Tasks.STATUS_COMPLETED ->  Status.VTODO_COMPLETED
            Tasks.STATUS_CANCELLED ->  Status.VTODO_CANCELLED
            else ->                    Status.VTODO_NEEDS_ACTION
        }

        val allDay = values.getAsInteger(Tasks.IS_ALLDAY) ?: 0 != 0

        val tzID = values.getAsString(Tasks.TZ)
        val tz = tzID?.let { DateUtils.ical4jTimeZone(it) }

        values.getAsLong(Tasks.CREATED)?.let { task.createdAt = it }
        values.getAsLong(Tasks.LAST_MODIFIED)?.let { task.lastModified = it }

        values.getAsLong(Tasks.DTSTART)?.let { dtStart ->
            task.dtStart =
                    if (allDay)
                        DtStart(Date(dtStart))
                    else {
                        val dt = DateTime(dtStart)
                        if (tz == null)
                            DtStart(dt, true)
                        else
                            DtStart(dt.apply {
                                timeZone = tz
                            })
                    }
        }

        values.getAsLong(Tasks.DUE)?.let { due ->
            task.due =
                    if (allDay)
                        Due(Date(due))
                    else {
                        val dt = DateTime(due)
                        if (tz == null)
                            Due(dt).apply {
                                isUtc = true
                            }
                        else
                            Due(dt.apply {
                                timeZone = tz
                            })
                    }
        }

        values.getAsString(Tasks.DURATION)?.let { duration ->
            val fixedDuration = AndroidTimeUtils.parseDuration(duration)
            task.duration = Duration(fixedDuration)
        }

        values.getAsString(Tasks.RDATE)?.let {
            task.rDates += AndroidTimeUtils.androidStringToRecurrenceSet(it, allDay) { dates -> RDate(dates) }
        }
        values.getAsString(Tasks.EXDATE)?.let {
            task.exDates += AndroidTimeUtils.androidStringToRecurrenceSet(it, allDay) { dates -> ExDate(dates) }
        }

        values.getAsString(Tasks.RRULE)?.let { task.rRule = RRule(it) }
    }

    protected open fun populateProperty(row: ContentValues) {
        Ical4Android.log.log(Level.FINER, "Found property", row)

        val task = requireNotNull(task)
        when (val type = row.getAsString(Properties.MIMETYPE)) {
            Alarm.CONTENT_ITEM_TYPE ->
                populateAlarm(row)
            Category.CONTENT_ITEM_TYPE ->
                task.categories += row.getAsString(Category.CATEGORY_NAME)
            Relation.CONTENT_ITEM_TYPE ->
                populateRelatedTo(row)
            UnknownProperty.CONTENT_ITEM_TYPE ->
                task.unknownProperties += UnknownProperty.fromJsonString(row.getAsString(UNKNOWN_PROPERTY_DATA))
            else ->
                Ical4Android.log.warning("Found unknown property of type $type")
        }
    }

    protected open fun populateAlarm(row: ContentValues) {
        val task = requireNotNull(task)
        val props = PropertyList<Property>()

        val trigger = Trigger(java.time.Duration.ofMinutes(-row.getAsLong(Alarm.MINUTES_BEFORE)))
        when (row.getAsInteger(Alarm.REFERENCE)) {
            Alarm.ALARM_REFERENCE_START_DATE ->
                trigger.parameters.add(Related.START)
            Alarm.ALARM_REFERENCE_DUE_DATE ->
                trigger.parameters.add(Related.END)
        }
        props += trigger

        props += when (row.getAsInteger(Alarm.ALARM_TYPE)) {
            Alarm.ALARM_TYPE_EMAIL ->
                Action.EMAIL
            Alarm.ALARM_TYPE_SOUND ->
                Action.AUDIO
            else ->
                // show alarm by default
                Action.DISPLAY
        }

        props += Description(row.getAsString(Alarm.MESSAGE) ?: task.summary)

        task.alarms += VAlarm(props)
    }

    protected open fun populateRelatedTo(row: ContentValues) {
        val uid = row.getAsString(Relation.RELATED_UID)
        if (uid == null) {
            Ical4Android.log.warning("Task relation doesn't refer to same task list; can't be synchronized")
            return
        }

        val relatedTo = RelatedTo(uid)

        // add relation type as reltypeparam
        relatedTo.parameters.add(when (row.getAsInteger(Relation.RELATED_TYPE)) {
            Relation.RELTYPE_CHILD ->
                RelType.CHILD
            Relation.RELTYPE_SIBLING ->
                RelType.SIBLING
            else /* Relation.RELTYPE_PARENT, default value */ ->
                RelType.PARENT
        })

        requireNotNull(task).relatedTo.add(relatedTo)
    }


    fun add(): Uri {
        val batch = BatchOperation(taskList.provider.client)

        val builder = CpoBuilder.newInsert(taskList.tasksSyncUri())
        buildTask(builder, false)
        val idxTask = batch.nextBackrefIdx()
        batch.enqueue(builder)

        insertProperties(batch, idxTask)

        batch.commit()

        val resultUri = batch.getResult(0)?.uri ?: throw CalendarStorageException("Empty result from provider when adding a task")
        id = ContentUris.parseId(resultUri)
        return resultUri
    }

    fun update(task: Task): Uri {
        this.task = task
        val existingId = requireNotNull(id)

        val batch = BatchOperation(taskList.provider.client)

        // remove associated rows which are added later again
        batch.enqueue(CpoBuilder
                .newDelete(taskList.tasksPropertiesSyncUri())
                .withSelection("${Properties.TASK_ID}=?", arrayOf(existingId.toString())))

        // update task
        val uri = taskSyncURI()
        val builder = CpoBuilder.newUpdate(uri)
        buildTask(builder, true)
        batch.enqueue(builder)

        // insert task properties again
        insertProperties(batch, null)

        batch.commit()
        return ContentUris.withAppendedId(taskList.provider.tasksUri(), existingId)
    }

    protected open fun insertProperties(batch: BatchOperation, idxTask: Int?) {
        insertAlarms(batch, idxTask)
        insertCategories(batch, idxTask)
        insertRelatedTo(batch, idxTask)
        insertUnknownProperties(batch, idxTask)
    }

    protected open fun insertAlarms(batch: BatchOperation, idxTask: Int?) {
        val task = requireNotNull(task)
        for (alarm in task.alarms) {
            val (alarmRef, minutes) = ICalendar.vAlarmToMin(alarm, task, true) ?: continue
            val ref = when (alarmRef) {
                Related.END ->
                    Alarm.ALARM_REFERENCE_DUE_DATE
                else /* Related.START is the default value */ ->
                    Alarm.ALARM_REFERENCE_START_DATE
            }

            val alarmType = when (alarm.action?.value?.uppercase(Locale.ROOT)) {
                Action.AUDIO.value ->
                    Alarm.ALARM_TYPE_SOUND
                Action.DISPLAY.value ->
                    Alarm.ALARM_TYPE_MESSAGE
                Action.EMAIL.value ->
                    Alarm.ALARM_TYPE_EMAIL
                else ->
                    Alarm.ALARM_TYPE_NOTHING
            }

            val builder = CpoBuilder
                    .newInsert(taskList.tasksPropertiesSyncUri())
                    .withTaskId(Alarm.TASK_ID, idxTask)
                    .withValue(Alarm.MIMETYPE, Alarm.CONTENT_ITEM_TYPE)
                    .withValue(Alarm.MINUTES_BEFORE, minutes)
                    .withValue(Alarm.REFERENCE, ref)
                    .withValue(Alarm.MESSAGE, alarm.description?.value ?: alarm.summary)
                    .withValue(Alarm.ALARM_TYPE, alarmType)

            Ical4Android.log.log(Level.FINE, "Inserting alarm", builder.build())
            batch.enqueue(builder)
        }
    }

    protected open fun insertCategories(batch: BatchOperation, idxTask: Int?) {
        for (category in requireNotNull(task).categories) {
            val builder = CpoBuilder.newInsert(taskList.tasksPropertiesSyncUri())
                    .withTaskId(Category.TASK_ID, idxTask)
                    .withValue(Category.MIMETYPE, Category.CONTENT_ITEM_TYPE)
                    .withValue(Category.CATEGORY_NAME, category)
            Ical4Android.log.log(Level.FINE, "Inserting category", builder.build())
            batch.enqueue(builder)
        }
    }

    protected open fun insertRelatedTo(batch: BatchOperation, idxTask: Int?) {
        for (relatedTo in requireNotNull(task).relatedTo) {
            val relType = when ((relatedTo.getParameter(Parameter.RELTYPE) as RelType?)) {
                RelType.CHILD ->
                    Relation.RELTYPE_CHILD
                RelType.SIBLING ->
                    Relation.RELTYPE_SIBLING
                else /* RelType.PARENT, default value */ ->
                    Relation.RELTYPE_PARENT
            }
            val builder = CpoBuilder.newInsert(taskList.tasksPropertiesSyncUri())
                    .withTaskId(Relation.TASK_ID, idxTask)
                    .withValue(Relation.MIMETYPE, Relation.CONTENT_ITEM_TYPE)
                    .withValue(Relation.RELATED_UID, relatedTo.value)
                    .withValue(Relation.RELATED_TYPE, relType)
            Ical4Android.log.log(Level.FINE, "Inserting relation", builder.build())
            batch.enqueue(builder)
        }
    }

    protected open fun insertUnknownProperties(batch: BatchOperation, idxTask: Int?) {
        for (property in requireNotNull(task).unknownProperties) {
            if (property.value.length > UnknownProperty.MAX_UNKNOWN_PROPERTY_SIZE) {
                Ical4Android.log.warning("Ignoring unknown property with ${property.value.length} octets (too long)")
                return
            }

            val builder = CpoBuilder.newInsert(taskList.tasksPropertiesSyncUri())
                    .withTaskId(Properties.TASK_ID, idxTask)
                    .withValue(Properties.MIMETYPE, UnknownProperty.CONTENT_ITEM_TYPE)
                    .withValue(UNKNOWN_PROPERTY_DATA, UnknownProperty.toJsonString(property))
            Ical4Android.log.log(Level.FINE, "Inserting unknown property", builder.build())
            batch.enqueue(builder)
        }
    }

    fun delete(): Int {
        try {
            return taskList.provider.client.delete(taskSyncURI(), null, null)
        } catch(e: RemoteException) {
            throw CalendarStorageException("Couldn't delete event", e)
        }
    }

    @CallSuper
    protected open fun buildTask(builder: CpoBuilder, update: Boolean) {
        if (!update)
            builder .withValue(Tasks.LIST_ID, taskList.id)

        val task = requireNotNull(task)
        builder .withValue(Tasks._UID, task.uid)
                .withValue(Tasks._DIRTY, 0)
                .withValue(Tasks.SYNC_VERSION, task.sequence)
                .withValue(Tasks.TITLE, task.summary)
                .withValue(Tasks.LOCATION, task.location)
                .withValue(Tasks.GEO, task.geoPosition?.let { "${it.longitude},${it.latitude}" })
                .withValue(Tasks.DESCRIPTION, task.description)
                .withValue(Tasks.TASK_COLOR, task.color)
                .withValue(Tasks.URL, task.url)

                // parent_id will be re-calculated when the relation row is inserted (if there is any)
                .withValue(Tasks.PARENT_ID, null)

        task.organizer?.let { organizer ->
            val uri = organizer.calAddress
            val email = if (uri.scheme.equals("mailto", true))
                uri.schemeSpecificPart
            else
                organizer.getParameter<Email>(Parameter.EMAIL)?.value
            if (email != null)
                builder.withValue(Tasks.ORGANIZER, email)
            else
                Ical4Android.log.warning("Ignoring ORGANIZER without email address (not supported by Android)")
        }

        builder .withValue(Tasks.PRIORITY, task.priority)
                .withValue(Tasks.CLASSIFICATION, when (task.classification) {
                    Clazz.PUBLIC -> Tasks.CLASSIFICATION_PUBLIC
                    Clazz.CONFIDENTIAL -> Tasks.CLASSIFICATION_CONFIDENTIAL
                    null -> Tasks.CLASSIFICATION_DEFAULT
                    else -> Tasks.CLASSIFICATION_PRIVATE    // all unknown classifications MUST be treated as PRIVATE
                })

        // COMPLETED must always be a DATE-TIME
        builder .withValue(Tasks.COMPLETED, task.completedAt?.date?.time)
                .withValue(Tasks.COMPLETED_IS_ALLDAY, 0)
                .withValue(Tasks.PERCENT_COMPLETE, task.percentComplete)

        val status = when (task.status) {
            Status.VTODO_IN_PROCESS -> Tasks.STATUS_IN_PROCESS
            Status.VTODO_COMPLETED  -> Tasks.STATUS_COMPLETED
            Status.VTODO_CANCELLED  -> Tasks.STATUS_CANCELLED
            else                    -> Tasks.STATUS_DEFAULT    // == Tasks.STATUS_NEEDS_ACTION
        }
        builder.withValue(Tasks.STATUS, status)

        val allDay = task.isAllDay()
        if (allDay) {
            builder .withValue(Tasks.IS_ALLDAY, 1)
                    .withValue(Tasks.TZ, null)
        } else {
            AndroidTimeUtils.androidifyTimeZone(task.dtStart)
            AndroidTimeUtils.androidifyTimeZone(task.due)
            builder .withValue(Tasks.IS_ALLDAY, 0)
                    .withValue(Tasks.TZ, getTimeZone().id)
        }

        builder .withValue(Tasks.CREATED, task.createdAt)
                .withValue(Tasks.LAST_MODIFIED, task.lastModified)

                .withValue(Tasks.DTSTART, task.dtStart?.date?.time)
                .withValue(Tasks.DUE, task.due?.date?.time)
                .withValue(Tasks.DURATION, task.duration?.value)

                .withValue(Tasks.RDATE,
                        if (task.rDates.isEmpty())
                            null
                        else
                            AndroidTimeUtils.recurrenceSetsToOpenTasksString(task.rDates, if (allDay) null else getTimeZone()))
                .withValue(Tasks.RRULE, task.rRule?.value)

                .withValue(Tasks.EXDATE,
                        if (task.exDates.isEmpty())
                            null
                        else
                            AndroidTimeUtils.recurrenceSetsToOpenTasksString(task.exDates, if (allDay) null else getTimeZone()))
        Ical4Android.log.log(Level.FINE, "Built task object", builder.build())
    }


    fun getTimeZone(): TimeZone {
        val task = requireNotNull(task)
        return  task.dtStart?.let { dtStart ->
                    if (dtStart.isUtc)
                        utcTimeZone
                    else
                        dtStart.timeZone
                } ?:
                task.due?.let { due ->
                    if (due.isUtc)
                        utcTimeZone
                    else
                        due.timeZone
                } ?:
                DateUtils.ical4jTimeZone(ZoneId.systemDefault().id)!!
    }


    protected fun CpoBuilder.withTaskId(column: String, idxTask: Int?): CpoBuilder {
        if (idxTask != null)
            withValueBackReference(column, idxTask)
        else
            withValue(column, requireNotNull(id))
        return this
    }


    protected fun taskSyncURI(loadProperties: Boolean = false): Uri {
        val id = requireNotNull(id)
        return ContentUris.withAppendedId(taskList.tasksSyncUri(loadProperties), id)
    }

    override fun toString() = MiscUtils.reflectionToString(this)

}
